StealFocus/steal-focus-explainer.md

```diff
@@ -0,0 +1,79 @@
+
+## Proposed solution: "disallowprogrammaticfocus" attribute
+We propose a new "disallowprogrammaticfocus" boolean attribute on the [HTMLIFrameElement](https://www.w3.org/TR/2011/WD-html5-20110525/the-iframe-element.html#the-iframe-element). Whenever it is set, all the nested iframes can no longer take input focus through script.
+
+```html
+<!DOCTYPE html> 
+<html> 
+  <head> 
+    <meta charset="utf-8"> 
+    <title>iframe steal focus prevention</title> 
+  </head> 
+  <body> 
+    <iframe id="extension-tab-frame"
+            class="extension-tab-frame"
+            name="extension-tab-frame"
+            src=""
+            frameborder="0"
+            sandbox="allow-forms allow-modals allow-popups allow-pointer-lock allow-scripts allow-same-origin allow-downloads"
+            allow="geolocation *; microphone *; camera *; midi *; encrypted-media *;"
+            allowfullscreen
+            disallowprogrammaticfocus>
+    </iframe>
+  </body> 
+</html> 
+```
+
+
+
+## Alternative Solutions
+This section exposes some of the alternative solutions that we came across before coming up with the chosen proposal.
+
+### Add "programmaticfocus" option to "allow" or "sandbox" attribute.
+Similar to the "media-playback-while-not-rendered" [Permission Policy](https://www.w3.org/TR/permissions-policy/) on the [HTMLIFrameElement](https://www.w3.org/TR/2011/WD-html5-20110525/the-iframe-element.html#the-iframe-element). We chould have a new policy "programmatic-focus". When set, the nest iframes could take input focus through script.
+
+```html
+<!DOCTYPE html> 
+<html> 
+  <head> 
+    <meta charset="utf-8"> 
+    <title>iframe steal focus prevention</title> 
+  </head> 
+  <body> 
+    <iframe id="extension-tab-frame"
+            class="extension-tab-frame"
+            name="extension-tab-frame"
+            src=""
+            frameborder="0"
+            sandbox="allow-programmatic-focus allow-forms allow-modals allow-popups allow-pointer-lock allow-scripts allow-same-origin allow-downloads"
+            allow="programmatic-focus *; geolocation *; microphone *; camera *; midi *; encrypted-media *;"
+            allowfullscreen>
+    </iframe>
+  </body> 
+</html> 
+```
+
+This alternative was not selected as the preferred one, because we think that the default option (policy not set) is against current behavior and may have compatibility risk.
```

Comment:
Potentially we could introduce a "disallow" attribute that can be used to turn off programmatic focus, leaving the default behavior the same as it is today. Would that address the compat risk? Is that approach better or worse than our currently proposed solution?

Reply:
There is already a "allow" attribute on iframe. I think it could potentially create confusions if we have a "disallow" because there will be three states: allow, disallow, unstated.  Therefore, I think the current proposed solution is better.